// picam/index.js
//PiCamStream utilizes the Raspberry Pi's raspistill program, monitoring the image it creates for changes and reporting those change events.

//Special thanks to Arvind Ravulavaru with camera functionality referenced from his work at:
//http://thejackalofjavascript.com/rpi-live-streaming/

//!!! TO DISCUSS:
//Should require() happen outside in global scope or inside object?
//In this case, both were used. Is this ok or just outside in module's global for the future?

//!!! expand on this module: raspistill has a lot of options and modes that can be utilized!
var events = require('events');

var PiCam = function() {
	//capture settings
	this.imgWidth = 640;
	this.imgHeight = 480;
	this.captureInterval = 1000; //in milliseconds
	this.checkFileInterval = 500; //events called when cam image file changes

	this.fs = require('fs');
	this.path = require('path');

	this.spawn = require('child_process').spawn;
	this.proc;

	this.active = false;

	//where the raspistill image is saved
	this.imgFile = this.path.join(__dirname, 'PiCam/PiCam.jpg')
};

//inherit all the properties from EventEmitter so that this object can emit its own events to start
PiCam.prototype = new events.EventEmitter;
 
//stop streaming camera and sensors
PiCam.prototype.stop = function() {
	var self = this;

	self.active = false;
	
	if(self.proc) {
		self.proc.kill();
	}
	
	self.fs.unwatchFile(self.imgFile);
};

PiCam.prototype.watchCamFile = function() {
	var self = this;

	self.fs.watchFile(self.imgFile, { persistent: true, interval: self.checkFileInterval }, function(current, previous) {
		self.emit('imageupdate', self.imgFile);
	});
};	
 
PiCam.prototype.start = function(io) {
 	var args, self = this;
	
 	//!!! currently set for upside down camera... create options for different orientations
	args = ["-vf", "-hf", "-w", this.imgWidth, "-h", this.imgHeight, "-o", self.imgFile, "-t", "999999999", "-tl", this.captureInterval];
	self.proc = self.spawn('raspistill', args);
 
	console.log('Watching PiCam image for changes...');
 
	self.active = true;
 
	self.fs.exists(self.imgFile, function(exists) {
		if(exists) {
			self.watchCamFile();
		} else {
			self.fs.closeSync(self.fs.openSync(self.imgFile, 'w'));
			self.watchCamFile();
		}
	});

};

module.exports = new PiCam();
