// picam/index.js
//PiCamStream utilizes the Raspberry Pi's raspistill program, monitoring the image it creates for changes and reporting those change events.

//Special thanks to Arvind Ravulavaru with camera functionality referenced from his work at:
//http://thejackalofjavascript.com/rpi-live-streaming/

//!!! expand on this module: raspistill has a lot of options and modes that can be utilized!

var events = require('events');
var fs = require('fs');
var path = require('path');
var spawn = require('child_process').spawn;

var PiCam = function() {
	//capture settings
	this.imgWidth = 640;
	this.imgHeight = 480;
	this.captureInterval = 1000; //in milliseconds
	this.checkFileInterval = 500; //events called when cam image file changes

	this.proc;

	this.active = false;

	//where the raspistill image is saved
	this.imgFile = path.join(__dirname, 'PiCam/PiCam.jpg')
};

//inherit all the properties from EventEmitter so that this object can emit its own events to start
PiCam.prototype = new events.EventEmitter;

PiCam.prototype.watchCamFile = function() {
	var self = this;

	fs.watchFile(self.imgFile, { persistent: true, interval: self.checkFileInterval }, function(current, previous) {
		self.emit('imageupdate', self.imgFile);
	});
};	
 
//stop streaming camera and sensors
PiCam.prototype.stop = function() {
	var self = this;

	self.active = false;
	
	if(self.proc) {
		console.log('killing raspistill process...');
		self.proc.kill();
	}
	
	//!!! need named callback referenced here as well I think, otherwise multiple events occur next time
	fs.unwatchFile(self.imgFile);
};
 
PiCam.prototype.start = function(io) {
 	var args, self = this;
	
 	//!!! currently set for upside down camera... create options for different orientations
	args = ["-vf", "-hf", "-w", this.imgWidth, "-h", this.imgHeight, "-o", self.imgFile, "-t", "999999999", "-tl", this.captureInterval];
	self.proc = spawn('raspistill', args);
 
	console.log('Watching PiCam image for changes...');
 
	self.active = true;
 
	fs.exists(self.imgFile, function(exists) {
		if(exists) {
			self.watchCamFile();
		} else {
			fs.closeSync(fs.openSync(self.imgFile, 'w'));
			self.watchCamFile();
		}
	});

};

module.exports = new PiCam();
